crates/core/tedge_api/src/alarm.rs

```diff
@@ -1,58 +1,84 @@
-use std::convert::TryFrom;
-use std::convert::TryInto;
-use std::fmt;
-
+use crate::entity_store::EntityMetadata;
 use clock::Timestamp;
 use serde::Deserialize;
 use serde::Serialize;
-
 use serde_json::Value;
 use std::collections::HashMap;
-/// In-memory representation of ThinEdge JSON alarm.
-#[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
+use std::convert::TryFrom;
+
+/// Internal representation of ThinEdge alarm model.
+#[derive(Debug, Eq, PartialEq)]
 pub struct ThinEdgeAlarm {
-    pub name: String,
-    pub severity: AlarmSeverity,
-    #[serde(flatten)]
-    pub data: Option<ThinEdgeAlarmData>,
-    pub source: Option<String>,
-}
+    /// Alarm type retrieved from the MQTT topic or payload.
+    /// The "type" given from the payload has priority to the one from the topic.
+    pub alarm_type: Option<String>,
 
-#[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
-pub enum AlarmSeverity {
-    Critical,
-    Major,
-    Minor,
-    Warning,
+    /// Alarm source retrieved from the MQTT topic.
+    pub source: EntityMetadata,
```

Comment:
I think it would be more appropriate to use an [`EntityTopicId`](https://github.com/thin-edge/thin-edge.io/blob/main/crates/core/tedge_api/src/mqtt_topics.rs#L213) instead of an [`EntityMetadata`](https://github.com/thin-edge/thin-edge.io/blob/main/crates/core/tedge_api/src/entity_store.rs#L288).

- An `EntityTopicId` is just an identifier such as `device/xyz//`, while an `EntityMetadata` contains all the data attached to that entity.
- Using an `EntityMetadata` to identify the source of an alarm makes things a bit more involved (one needs an [`EntityStore`](https://github.com/thin-edge/thin-edge.io/blob/main/crates/core/tedge_api/src/entity_store.rs#L58) to decode the `EntityTopicId` extracted from the alarm topic.
- One can not represent an alarm sent by a device that is not registered yet (as it is then unknown from the `EntityStore`).